### PR TITLE
[models] added baichuan/baichuan2 model support.

### PR DESCRIPTION
**Pull request by liutongxuan**

### CODE REVIEW

File: src/models/huggingface/baichuan.h [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+  }
+
+  torch::Tensor forward(torch::Tensor x,
+                        torch::Tensor positions,
+                        torch::Tensor* residual,
+                        KVCache& kv_cache,
+                        const InputParameters& input_params) {
+    torch::Tensor hidden_states;
+    if (residual == nullptr) {
+      residual = new torch::Tensor();
```
Comment (guocuimi) [inline src/models/huggingface/baichuan.h:287]: sounds you forget to release the memory newed here.
in pytorch, a Tensor is essentially a shared ptr already. if we follow the python style, the code can be rewerite in this way.

```
pass in torch::Tensor& residual

if (!residual.defined()) {
    residual = x;
}

```

File: src/layers/normalization.h [hunk 2]
```diff
@@ -172,4 +172,67 @@ class RMSNormImpl : public torch::nn::Module {
...
+              torch::ScalarType dtype,
+              const torch::Device& device)
+      : eps_(eps) {
+    weight_ = register_parameter(
+        "weight",
+        torch::empty({dim}, torch::dtype(dtype).device(device)),
+        /*requires_grad=*/false);
+  }
+
+  torch::Tensor forward(torch::Tensor input, torch::Tensor* residual) {
```
Comment (guocuimi) [inline src/layers/normalization.h:189]: Given that torch::Tensor functions somewhat like a shared pointer, adhering to the PyTorch convention, we can use torch::Tensor& for the out parameter. While this style may not be optimal and could obscure the purpose of in and out parameters, it aligns with PyTorch's conventions. Typically, pointers are preferred for out parameters, but for consistency, let's adhere to PyTorch's style for now.

File: src/layers/normalization.h [hunk 3]
```diff
@@ -172,4 +172,67 @@ class RMSNormImpl : public torch::nn::Module {
...
+    if (input.is_cuda() && !FLAGS_disable_custom_kernels) {
+      auto output = torch::empty_like(input);
+      if (residual != nullptr) {
+        input = input + *residual;
+        *residual = input;
+      }
+      kernel::rms_norm(output, input, weight_, eps_);
+      return output;
+    }
+    if (residual != nullptr) {
```
Comment (guocuimi) [inline src/layers/normalization.h:199]: It appears that introducing this class with a residual might be optional since this logic could be entirely moved out of the function. For instance:
```
if (residual.defined()) {
    input = input + residual;
    residual = input;
}
rms_norm(input);
```

but it may be necessary to be a new class if we want to fuse the kernel in the future.

**Comment by guocuimi** [comment]
LGTM in general. let me know once it passed correctness test. thanks

File: src/models/huggingface/baichuan.h (hunk 1 above)
Comment (liutongxuan) [reply src/models/huggingface/baichuan.h:287]: good catch

**Comment by liutongxuan** [comment]
> sounds the model weights was not packed in pickle format. i remember that i encounter this issue when supporting other models as well, which is tracked by #9. To workaround, we may need a python tool to convert the weight to pickle or safetensors format.
> 
> ```
> frame #4: torch::jit::Unpickler::readInstruction() + 0x1357 (0x7f3d9977d437 in /home/michael/.local/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
> frame #5: torch::jit::Unpickler::run() + 0x90 (0x7f3d9977f080 in /home/michael/.local/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
> frame #6: torch::jit::Unpickler::parse_ivalue() + 0x18 (0x7f3d997809d8 in /home/michael/.local/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
> frame #7: torch::jit::readArchiveAndTensors(std::string const&, std::string const&, std::string const&, c10::optional<std::function<c10::StrongTypePtr (c10::QualifiedName const&)> >,
> ```

yep, i have tried and it works.

File: src/layers/normalization.h (hunk 3 above)
Comment (liutongxuan) [reply src/layers/normalization.h:199]: yes, a new class with fused kernel could be better.
